// graphql-executor/type/directives.mjs
import {
  DirectiveLocation,
  GraphQLBoolean,
  GraphQLDirective,
  GraphQLInt,
  GraphQLString,
} from 'graphql';
/**
 * Used to conditionally defer fragments.
 */

export const GraphQLDeferDirective = new GraphQLDirective({
  name: 'defer',
  description:
    'Directs the executor to defer this fragment when the `if` argument is true or undefined.',
  locations: [
    DirectiveLocation.FRAGMENT_SPREAD,
    DirectiveLocation.INLINE_FRAGMENT,
  ],
  args: {
    if: {
      type: GraphQLBoolean,
      description: 'Deferred when true or undefined.',
    },
    label: {
      type: GraphQLString,
      description: 'Unique name',
    },
  },
});
/**
 * Used to conditionally stream list fields.
 */

export const GraphQLStreamDirective = new GraphQLDirective({
  name: 'stream',
  description:
    'Directs the executor to stream plural fields when the `if` argument is true or undefined.',
  locations: [DirectiveLocation.FIELD],
  args: {
    if: {
      type: GraphQLBoolean,
      description: 'Stream when true or undefined.',
    },
    label: {
      type: GraphQLString,
      description: 'Unique name',
    },
    initialCount: {
      defaultValue: 0,
      type: GraphQLInt,
      description: 'Number of items to return immediately',
    },
    maxChunkSize: {
      defaultValue: 1,
      type: GraphQLInt,
      description: 'Maximum number of items to return within each payload',
    },
    maxInterval: {
      type: GraphQLInt,
      description:
        'Maximum time in ms to wait to collect items for each payload, will wait indefinitely if undefined',
    },
    inParallel: {
      defaultValue: false,
      type: GraphQLBoolean,
      description: 'Stream items non-sequentially on completion if true',
    },
  },
});
